dd-sdk-android/src/main/kotlin/com/datadog/android/v2/core/internal/storage/ConsentAwareStorage.kt

```diff
@@ -39,56 +42,46 @@ internal class ConsentAwareStorage(
     @WorkerThread
     override fun writeCurrentBatch(
         datadogContext: DatadogContext,
-        callback: (BatchWriter) -> Unit
+        callback: (EventBatchWriter) -> Unit
     ) {
         val orchestrator = when (datadogContext.trackingConsent) {
             TrackingConsent.GRANTED -> grantedOrchestrator
             TrackingConsent.PENDING -> pendingOrchestrator
             TrackingConsent.NOT_GRANTED -> null
         }
 
-        val batchFile = orchestrator?.getWritableFile()
-        val metadataFile = if (batchFile != null) orchestrator.getMetadataFile(batchFile) else null
-
-        val writer = object : BatchWriter {
-            @WorkerThread
-            override fun currentMetadata(): ByteArray? {
-                if (orchestrator == null || metadataFile == null) return null
-
-                return batchMetadataReaderWriter.readData(metadataFile)
-            }
-
-            @WorkerThread
-            override fun write(
-                event: ByteArray,
-                eventId: String,
-                newMetadata: ByteArray?,
-                listener: BatchWriterListener
-            ) {
-                // prevent useless operation for empty event / null orchestrator
-                if (event.isEmpty() || orchestrator == null) {
-                    listener.onDataWritten(eventId)
-                    return
-                }
-
-                if (!checkEventSize(event.size)) {
-                    listener.onDataWriteFailed(eventId)
-                    return
+        try {
+            @Suppress("UnsafeThirdPartyFunctionCall") // command is never null
+            executor.execute {
+                val batchFile = orchestrator?.getWritableFile()
+                val metadataFile = if (batchFile != null) {
+                    orchestrator.getMetadataFile(batchFile)
+                } else {
+                    null
                 }
-
-                if (batchFile != null &&
-                    batchEventsReaderWriter.writeData(batchFile, event, true)
-                ) {
-                    if (newMetadata?.isNotEmpty() == true && metadataFile != null) {
-                        writeBatchMetadata(metadataFile, newMetadata)
-                    }
-                    listener.onDataWritten(eventId)
+                val writer = if (orchestrator == null || batchFile == null) {
+                    NoOpEventBatchWriter()
                 } else {
-                    listener.onDataWriteFailed(eventId)
+                    FileEventBatchWriter(
```

Comment:
are we creating a writer every time we are writing something ?

Reply:
yes, because writer is associated with a particular batch file

Reply:
Damn...I am thinking that for SR I will have to write tons of events....meaning that we will be creating those writers all the time. Aren't we afraid of too many objects there ? I mean I know they are not so heavy but was thinking that maybe if we could do it differently 🤔

Reply:
We can profile it. Anyway, the number of these objects will be much less that number of snapshot/wireframes objects to write :)